### ISSUE TITLE
GUI HA doesn't work in case of network policy is present only for one GUI node

### ISSUE THREAD

**Issue by saurabhwani5** [IBM/ibm-spectrum-scale-csi#1055]
## Describe the bug
There are 3 GUIs present in k8s cluster and the network policy for the GUI which is being used not present or delete by the user then if try to create PVC , it gives error as `Warning  ProvisioningFailed    92s (x8 over 10m)    spectrumscale.csi.ibm.com_ibm-spectrum-scale-csi-provisioner-89bf76648-85b4c_0dc1d376-3b5e-44a3-9f37-759f9c6ad4c2  failed to provision volume with StorageClass "ibm-spectrum-scale-csi-advance": rpc error: code = Internal desc = Error in authentication: GET request [https://10.11.70.196:443/scalemgmt/v2/filesystems/fs1](https://10.11.70.196/scalemgmt/v2/filesystems/fs1), user: csiadmin, param: <nil>, response: <nil>` and request doesn't go to other GUI's even if network policy is present on other GUIs.

## How to Reproduce?

1. Install CSI with #1050 images 
```
[root@multiguiubu-master pr1050]# oc get pods
NAME                                                  READY   STATUS    RESTARTS   AGE
ibm-spectrum-scale-csi-4v5qv                          3/3     Running   0          27m
ibm-spectrum-scale-csi-7gvj6                          3/3     Running   0          27m
ibm-spectrum-scale-csi-attacher-6ffd46f894-cq9vd      1/1     Running   0          27m
ibm-spectrum-scale-csi-attacher-6ffd46f894-np9ql      1/1     Running   0          27m
ibm-spectrum-scale-csi-operator-6bbbb8c5f5-qs5gn      1/1     Running   0          27m
ibm-spectrum-scale-csi-provisioner-89bf76648-85b4c    1/1     Running   0          27m
ibm-spectrum-scale-csi-resizer-76dbb644b6-24pnn       1/1     Running   0          27m
ibm-spectrum-scale-csi-snapshotter-797c96c5f9-2ml4w   1/1     Running   0          27m
[root@multiguiubu-master pr1050]# oc get cso
NAME                     VERSION   SUCCESS
ibm-spectrum-scale-csi   2.10.0    True
[root@multiguiubu-master pr1050]# oc describe pod | grep quay
    Image:         quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver:v2.10.0-021123
    Image ID:      quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver@sha256:0cb6a9ae9cc6930a850c492501f6bf9ca95e8214173fa88001d053ec223a3409
  Normal   Pulled     27m                 kubelet            Container image "quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver:v2.10.0-021123" already present on machine
    Image:         quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver:v2.10.0-021123
    Image ID:      quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver@sha256:0cb6a9ae9cc6930a850c492501f6bf9ca95e8214173fa88001d053ec223a3409
  Normal   Pulled     27m                 kubelet            Container image "quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver:v2.10.0-021123" already present on machine
    Image:         quay.io/hemalatha_gajendran/tunable_host_network_latest
    Image ID:      quay.io/hemalatha_gajendran/tunable_host_network_latest@sha256:1e25ff135f6b7ac9cf4d845d2341c5403f210f90014ce69fc9bbb2d799812cb5
      CSI_DRIVER_IMAGE:      quay.io/ibm-spectrum-scale-dev/ibm-spectrum-scale-csi-driver:v2.10.0-021123
  Normal  Pulled     27m   kubelet            Container image "quay.io/hemalatha_gajendran/tunable_host_network_latest" already present on machine
```

2. Create configmap as following and apply :
```
[root@multiguiubu-master pr1050]# cat cm.yaml
kind: ConfigMap
apiVersion: v1
metadata:
  name: ibm-spectrum-scale-csi-config
  namespace: ibm-spectrum-scale-csi-driver
data:
  HOST_NETWORK: DISABLED
  VAR_DRIVER_LOGLEVEL: TRACE
```
3. Create network policies for dns , api server and for all GUI nodes:
```
[root@multiguiubu-master network]# cat allow-dns-access.yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-dns-access
  namespace: ibm-spectrum-scale-csi-driver
spec:
  podSelector:
    matchLabels: {}
  policyTypes:
  - Egress
  - Ingress
  ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: kube-system
      podSelector:
        matchLabels:
          k8s-app: kube-dns
    ports:
    - protocol: UDP
      port: 53
    - protocol: TCP
      port: 53
    - protocol: TCP
      port: 9153
  egress:
  - to:
    - namespaceSelector:
        matchLabels:
          kubernetes.io/metadata.name: kube-system
      podSelector:
        matchLabels:
          k8s-app: kube-dns
    ports:
    - protocol: UDP
      port: 53
    - protocol: TCP
      port: 53
    - protocol: TCP
      port: 9153

[root@multiguiubu-master network]# kubectl get endpoints kubernetes -n default
NAME         ENDPOINTS          AGE
kubernetes   10.11.67.44:6443   220d
[root@multiguiubu-master network]# cat allow-egress-apiserver.yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-egress-apiserver
  namespace: ibm-spectrum-scale-csi-driver
spec:
  podSelector: {}
  policyTypes:
    - Egress
  egress:
    - to:
        - ipBlock:
            cidr: 10.11.67.44/32
      ports:
        - protocol: TCP
          port: 6443
        - protocol: TCP
          port: 443
[root@multiguiubu-master network]# cat allow-egress-gui-1.yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-egress-gui-1
  namespace: ibm-spectrum-scale-csi-driver
spec:
  podSelector: {}
  policyTypes:
    - Egress
  egress:
    - to:
        - ipBlock:
            cidr: 10.11.70.196/32
      ports:
        - protocol: TCP
          port: 443
[root@multiguiubu-master network]# cat allow-egress-gui-2.yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-egress-gui-2
  namespace: ibm-spectrum-scale-csi-driver
spec:
  podSelector: {}
  policyTypes:
    - Egress
  egress:
    - to:
        - ipBlock:
            cidr: 10.11.70.225/32
      ports:
        - protocol: TCP
          port: 443
[root@multiguiubu-master network]# cat allow-egress-gui-3.yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-egress-gui-3
  namespace: ibm-spectrum-scale-csi-driver
spec:
  podSelector: {}
  policyTypes:
    - Egress
  egress:
    - to:
        - ipBlock:
            cidr: 10.11.70.240/32
      ports:
        - protocol: TCP
          port: 443
```
4. Create PVC and from the logs determine the GUI which is being used :
```
[root@multiguiubu-master pr1050]# cat pvc.yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: scale-advance-pvc-5
spec:
  accessModes:
  - ReadWriteMany
  resources:
    requests:
      storage: 1Gi
  storageClassName: ibm-spectrum-scale-csi-advance

---

apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
   name: ibm-spectrum-scale-csi-advance
provisioner: spectrumscale.csi.ibm.com
parameters:
   volBackendFs: "fs1"
   version: "2"
reclaimPolicy: Delete
[root@multiguiubu-master pr1050]# oc apply -f pvc.yaml
persistentvolumeclaim/scale-advance-pvc-5 created
storageclass.storage.k8s.io/ibm-spectrum-scale-csi-advance created
[root@multiguiubu-master pr1050]# oc get pvc -w
NAME                  STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS                     AGE
scale-advance-pvc-5   Pending                                      ibm-spectrum-scale-csi-advance   5s
scale-advance-pvc-5   Pending   pvc-aaf362d2-2125-48f2-a14c-10953cfc6f7b   0                         ibm-spectrum-scale-csi-advance   17s
scale-advance-pvc-5   Bound     pvc-aaf362d2-2125-48f2-a14c-10953cfc6f7b   1Gi        RWX            ibm-spectrum-scale-csi-advance   17s
```
5. Delete the network policy of that GUI node:
```
From Driver pod logs we can find out that 10.11.70.196 GUI is getting used currently, so deleting the network policy of the same
[root@multiguiubu-master network]# cat allow-egress-gui-1.yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-egress-gui-1
  namespace: ibm-spectrum-scale-csi-driver
spec:
  podSelector: {}
  policyTypes:
    - Egress
  egress:
    - to:
        - ipBlock:
            cidr: 10.11.70.196/32
      ports:
        - protocol: TCP
          port: 443
[root@multiguiubu-master network]# oc delete -f allow-egress-gui-1.yaml
networkpolicy.networking.k8s.io "allow-egress-gui-1" deleted
[root@multiguiubu-master pr1050]# oc delete -f pvc.yaml
persistentvolumeclaim "scale-advance-pvc-5" deleted
storageclass.storage.k8s.io "ibm-spectrum-scale-csi-advance" deleted
```
6. Create PVC again :
```
[root@multiguiubu-master pr1050]# oc apply -f pvc.yaml
persistentvolumeclaim/scale-advance-pvc-5 created
storageclass.storage.k8s.io/ibm-spectrum-scale-csi-advance created
[root@multiguiubu-master pr1050]# oc get pvc -w
NAME                  STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS                     AGE
scale-advance-pvc-5   Pending                                      ibm-spectrum-scale-csi-advance   2s
scale-advance-pvc-5   Pending                                      ibm-spectrum-scale-csi-advance   9s
```
7. Check PVC description 
```
[root@multiguiubu-master pr1050]# oc get networkpolicy
NAME                     POD-SELECTOR   AGE
allow-dns-access         <none>         12m
allow-egress-apiserver   <none>         100m
allow-egress-gui-2       <none>         65m
allow-egress-gui-3       <none>         64m
[root@multiguiubu-master pr1050]# oc describe pvc
Name:          scale-advance-pvc-5
Namespace:     ibm-spectrum-scale-csi-driver
StorageClass:  ibm-spectrum-scale-csi-advance
Status:        Pending
Volume:
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: spectrumscale.csi.ibm.com
               volume.kubernetes.io/storage-provisioner: spectrumscale.csi.ibm.com
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:
Access Modes:
VolumeMode:    Filesystem
Used By:       <none>
Events:
  Type     Reason                Age               From                                                                                                               Message
  ----     ------                ----              ----                                                                                                               -------
  Warning  ProvisioningFailed    76s               persistentvolume-controller                                                                                        storageclass.storage.k8s.io "ibm-spectrum-scale-csi-advance" not found
  Normal   ExternalProvisioning  7s (x5 over 67s)  persistentvolume-controller                                                                                        waiting for a volume to be created, either by external provisioner "spectrumscale.csi.ibm.com" or manually created by system administrator
  Warning  ProvisioningFailed    7s                spectrumscale.csi.ibm.com_ibm-spectrum-scale-csi-provisioner-89bf76648-85b4c_0dc1d376-3b5e-44a3-9f37-759f9c6ad4c2  failed to provision volume with StorageClass "ibm-spectrum-scale-csi-advance": rpc error: code = Internal desc = Error in authentication: GET request https://10.11.70.196:443/scalemgmt/v2/filesystems/fs1, user: csiadmin, param: <nil>, response: <nil>
  Normal   Provisioning          6s (x2 over 67s)  spectrumscale.csi.ibm.com_ibm-spectrum-scale-csi-provisioner-89bf76648-85b4c_0dc1d376-3b5e-44a3-9f37-759f9c6ad4c2  External provisioner is provisioning volume for claim "ibm-spectrum-scale-csi-driver/scale-advance-pvc-5"
```

## Expected behavior
PVC creation should happen through other GUIs having  network policy 

### Data Collection and Debugging

/scale-csi/D.1055
csisnap.tar.gz

**Comment by saurabhwani5** [comment]
Another issue is seen when PVC is already created and we try to delete it when network policy is not preset for GUI node , In this case PVC gets deleted but pv and fileset stays there with following error:
```
[root@multiguiubu-master network]# oc logs ibm-spectrum-scale-csi-4v5qv | grep 399b0b1d-4f17-4b99-9785-d291e9d34fab
I1103 11:13:45.451027       1 utils.go:61] [399b0b1d-4f17-4b99-9785-d291e9d34fab] GRPC call: /csi.v1.Controller/DeleteVolume
I1103 11:13:45.451058       1 utils.go:64] [399b0b1d-4f17-4b99-9785-d291e9d34fab] GRPC request: volume_id:"1;1;13433156642562752464;31430B0A:6421C0A5;ab27b1c6-f26e-4f62-9179-98f353052bd6-ibm-spectrum-scale-csi-driver;pvc-5460424f-3ec1-459e-8aa8-ad569f7ff412;/ibm/fs1/ab27b1c6-f26e-4f62-9179-98f353052bd6-ibm-spectrum-scale-csi-driver/pvc-5460424f-3ec1-459e-8aa8-ad569f7ff412"
I1103 11:13:45.451102       1 controllerserver.go:1709] [399b0b1d-4f17-4b99-9785-d291e9d34fab] DeleteVolume [volume_id:"1;1;13433156642562752464;31430B0A:6421C0A5;ab27b1c6-f26e-4f62-9179-98f353052bd6-ibm-spectrum-scale-csi-driver;pvc-5460424f-3ec1-459e-8aa8-ad569f7ff412;/ibm/fs1/ab27b1c6-f26e-4f62-9179-98f353052bd6-ibm-spectrum-scale-csi-driver/pvc-5460424f-3ec1-459e-8aa8-ad569f7ff412" ]
I1103 11:13:45.451129       1 gpfs.go:185] [399b0b1d-4f17-4b99-9785-d291e9d34fab] ValidateControllerServiceRequest
I1103 11:13:45.451141       1 controllerserver.go:1728] [399b0b1d-4f17-4b99-9785-d291e9d34fab] Volume Id Members [{13433156642562752464 31430B0A:6421C0A5   pvc-5460424f-3ec1-459e-8aa8-ad569f7ff412  /ibm/fs1/ab27b1c6-f26e-4f62-9179-98f353052bd6-ibm-spectrum-scale-csi-driver/pvc-5460424f-3ec1-459e-8aa8-ad569f7ff412 true 1 ab27b1c6-f26e-4f62-9179-98f353052bd6-ibm-spectrum-scale-csi-driver 1}]
I1103 11:13:45.451157       1 rest_v2.go:1153] [399b0b1d-4f17-4b99-9785-d291e9d34fab] rest_v2 GetFilesystemName. UUID: 31430B0A:6421C0A5
I1103 11:13:45.451168       1 rest_v2.go:1035] [399b0b1d-4f17-4b99-9785-d291e9d34fab] rest_v2 doHTTP: urlSuffix: scalemgmt/v2/filesystems?filter=uuid=31430B0A:6421C0A5, method: GET, param: <nil>
I1103 11:13:45.451175       1 rest_v2.go:1037] [399b0b1d-4f17-4b99-9785-d291e9d34fab] rest_v2 doHTTP: endpoint: https://10.11.70.196:443/
I1103 11:13:45.451184       1 http_utils.go:58] [399b0b1d-4f17-4b99-9785-d291e9d34fab] http_utils HttpExecuteUserAuth. type: GET, url: https://10.11.70.196:443/scalemgmt/v2/filesystems?filter=uuid=31430B0A:6421C0A5, user: csiadmin
I1103 11:13:45.451190       1 http_utils.go:59] [399b0b1d-4f17-4b99-9785-d291e9d34fab] http_utils HttpExecuteUserAuth. request payload: <nil>
I1103 11:13:45.451217       1 http_utils.go:81] [399b0b1d-4f17-4b99-9785-d291e9d34fab] http_utils HttpExecuteUserAuth request: &{Method:GET URL:https://10.11.70.196:443/scalemgmt/v2/filesystems?filter=uuid=31430B0A:6421C0A5 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Accept:[application/json] Authorization:[Basic Y3NpYWRtaW46YWRtaW51c2Vy] Content-Type:[application/json]] Body:{Reader:null} GetBody:0x6a8860 ContentLength:4 TransferEncoding:[] Close:false Host:10.11.70.196:443 Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil> Cancel:<nil> Response:<nil> ctx:0xc000112000}
E1103 11:14:45.451566       1 rest_v2.go:1062] [399b0b1d-4f17-4b99-9785-d291e9d34fab] rest_v2 doHTTP: Error in authentication request on endpoint https://10.11.70.196:443/: Get "https://10.11.70.196:443/scalemgmt/v2/filesystems?filter=uuid=31430B0A:6421C0A5": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
E1103 11:14:45.451663       1 rest_v2.go:1160] [399b0b1d-4f17-4b99-9785-d291e9d34fab] Unable to get filesystem name for uuid 31430B0A:6421C0A5: rpc error: code = Internal desc = Error in authentication: GET request https://10.11.70.196:443/scalemgmt/v2/filesystems?filter=uuid=31430B0A:6421C0A5, user: csiadmin, param: <nil>, response: <nil>
E1103 11:14:45.451700       1 utils.go:69] [399b0b1d-4f17-4b99-9785-d291e9d34fab] GRPC error: rpc error: code = Internal desc = unable to get filesystem Name for Id [31430B0A:6421C0A5] and clusterId [13433156642562752464]. Error [rpc error: code = Internal desc = Error in authentication: GET request https://10.11.70.196:443/scalemgmt/v2/filesystems?filter=uuid=31430B0A:6421C0A5, user: csiadmin, param: <nil>, response: <nil>]
I1103 11:14:45.451729       1 utils.go:79] [399b0b1d-4f17-4b99-9785-d291e9d34fab] Time taken to execute /csi.v1.Controller/DeleteVolume request(in milliseconds): 60000
```

**Comment by amdabhad** [comment]
Fixed in https://github.com/IBM/ibm-spectrum-scale-csi/pull/1060, please verify

**Comment by saurabhwani5** [comment]
verified , thank @hemalathagajendran for fix #1060
